File: src/db-toolkit/operations/data_editor.py
"""Production-ready data editing operations."""

from typing import Dict, Any, Optional, List
from connectors.factory import ConnectorFactory
from core.models import DatabaseConnection
from utils.logger import logger


class DataEditor:
    """Handles safe data editing with validation and constraints."""
    
    async def update_row(
        self,
        connection: DatabaseConnection,
        table: str,
        schema_name: str,
        primary_key: Dict[str, Any],
        changes: Dict[str, Any]
    ) -> Dict[str, Any]:
        """Update a single row with validation."""
        if not primary_key:
            return {"success": False, "error": "Primary key required for update"}
        
        if not changes:
            return {"success": False, "error": "No changes provided"}
        
        try:
            connector = ConnectorFactory.create_connector(connection.db_type)
            await connector.connect(connection)
            
            # Build UPDATE query
            if connection.db_type.value == "mongodb":
                result = await self._update_mongodb(connector, table, primary_key, changes)
            else:
                result = await self._update_sql(connector, table, schema_name, primary_key, changes)
            
            await connector.disconnect()
            return result
            
        except Exception as e:
            logger.error(f"Update row failed on '{connection.name}.{table}': {str(e)}")
            return {"success": False, "error": str(e)}
    
    async def insert_row(
        self,
        connection: DatabaseConnection,
        table: str,
        schema_name: str,
        data: Dict[str, Any]
    ) -> Dict[str, Any]:
        """Insert a new row."""
        if not data:
            return {"success": False, "error": "No data provided"}
        
        try:
            connector = ConnectorFactory.create_connector(connection.db_type)
            await connector.connect(connection)
            
            if connection.db_type.value == "mongodb":
                result = await self._insert_mongodb(connector, table, data)
            else:
                result = await self._insert_sql(connector, table, schema_name, data)
            
            await connector.disconnect()
            return result
            
        except Exception as e:
            logger.error(f"Insert row failed on '{connection.name}.{table}': {str(e)}")
            return {"success": False, "error": str(e)}
    
    async def delete_row(
        self,
        connection: DatabaseConnection,
        table: str,
        schema_name: str,
        primary_key: Dict[str, Any]
    ) -> Dict[str, Any]:
        """Delete a row."""
        if not primary_key:
            return {"success": False, "error": "Primary key required for delete"}
        
        try:
            connector = ConnectorFactory.create_connector(connection.db_type)
            await connector.connect(connection)
            
            if connection.db_type.value == "mongodb":
                result = await self._delete_mongodb(connector, table, primary_key)
            else:
                result = await self._delete_sql(connector, table, schema_name, primary_key)
            
            await connector.disconnect()
            return result
            
        except Exception as e:
            logger.error(f"Delete row failed on '{connection.name}.{table}': {str(e)}")
            return {"success": False, "error": str(e)}
    
    async def _update_sql(
        self,
        connector,
        table: str,
        schema: str,
        primary_key: Dict[str, Any],
        changes: Dict[str, Any]
    ) -> Dict[str, Any]:
        """Update SQL row."""
        set_clause = ", ".join([f"{col} = '{val}'" for col, val in changes.items()])
        where_clause = " AND ".join([f"{col} = '{val}'" for col, val in primary_key.items()])
        
        if schema and schema != "main":
            query = f"UPDATE {schema}.{table} SET {set_clause} WHERE {where_clause}"
        else:
            query = f"UPDATE {table} SET {set_clause} WHERE {where_clause}"
        
        result = await connector.execute_query(query)
        
        if result.get("success"):
            return {"success": True, "message": "Row updated successfully"}
        else:
            return {"success": False, "error": result.get("error", "Update failed")}
    
    async def _insert_sql(
        self,
        connector,
        table: str,
        schema: str,
        data: Dict[str, Any]
    ) -> Dict[str, Any]:
        """Insert SQL row."""
        columns = ", ".join(data.keys())
        values = ", ".join([f"'{val}'" for val in data.values()])
        
        if schema and schema != "main":
            query = f"INSERT INTO {schema}.{table} ({columns}) VALUES ({values})"
        else:
            query = f"INSERT INTO {table} ({columns}) VALUES ({values})"
        
        result = await connector.execute_query(query)
        
        if result.get("success"):
            return {"success": True, "message": "Row inserted successfully"}
        else:
            return {"success": False, "error": result.get("error", "Insert failed")}
    
    async def _delete_sql(
        self,
        connector,
        table: str,
        schema: str,
        primary_key: Dict[str, Any]
    ) -> Dict[str, Any]:
        """Delete SQL row."""
        where_clause = " AND ".join([f"{col} = '{val}'" for col, val in primary_key.items()])
        
        if schema and schema != "main":
            query = f"DELETE FROM {schema}.{table} WHERE {where_clause}"
        else:
            query = f"DELETE FROM {table} WHERE {where_clause}"
        
        result = await connector.execute_query(query)
        
        if result.get("success"):
            return {"success": True, "message": "Row deleted successfully"}
        else:
            return {"success": False, "error": result.get("error", "Delete failed")}
    
    async def _update_mongodb(
        self,
        connector,
        collection: str,
        filter_doc: Dict[str, Any],
        changes: Dict[str, Any]
    ) -> Dict[str, Any]:
        """Update MongoDB document."""
        try:
            db = connector.connection.get_default_database()
            coll = db[collection]
            result = await coll.update_one(filter_doc, {"$set": changes})
            
            if result.modified_count > 0:
                return {"success": True, "message": "Document updated successfully"}
            else:
                return {"success": False, "error": "No document matched the filter"}
        except Exception as e:
            return {"success": False, "error": str(e)}
    
    async def _insert_mongodb(
        self,
        connector,
        collection: str,
        data: Dict[str, Any]
    ) -> Dict[str, Any]:
        """Insert MongoDB document."""
        try:
            db = connector.connection.get_default_database()
            coll = db[collection]
            result = await coll.insert_one(data)
            
            return {"success": True, "message": "Document inserted successfully", "id": str(result.inserted_id)}
        except Exception as e:
            return {"success": False, "error": str(e)}
    
    async def _delete_mongodb(
        self,
        connector,
        collection: str,
        filter_doc: Dict[str, Any]
    ) -> Dict[str, Any]:
        """Delete MongoDB document."""
        try:
            db = connector.connection.get_default_database()
            coll = db[collection]
            result = await coll.delete_one(filter_doc)
            
            if result.deleted_count > 0:
                return {"success": True, "message": "Document deleted successfully"}
            else:
                return {"success": False, "error": "No document matched the filter"}
        except Exception as e:
            return {"success": False, "error": str(e)}

File: src/db-toolkit/electron/backend/operations/data-editor.js
/**
 * Data editing operations.
 */

const { ConnectorFactory } = require('../connectors');
const { logger } = require('../utils/logger.js');

class DataEditor {
  async updateRow(connection, table, schemaName, primaryKey, changes) {
    if (!primaryKey || Object.keys(primaryKey).length === 0) {
      return { success: false, error: 'Primary key required for update' };
    }

    if (!changes || Object.keys(changes).length === 0) {
      return { success: false, error: 'No changes provided' };
    }

    try {
      const connector = ConnectorFactory.createConnector(connection.db_type);
      await connector.connect(connection);

      let result;
      if (connection.db_type === 'mongodb') {
        result = await this.updateMongoDB(connector, table, primaryKey, changes);
      } else {
        result = await this.updateSQL(connector, table, schemaName, primaryKey, changes);
      }

      await connector.disconnect();
      return result;
    } catch (error) {
      logger.error(`Update row failed on '${connection.name}.${table}':`, error);
      return { success: false, error: error.message };
    }
  }

  async insertRow(connection, table, schemaName, data) {
    if (!data || Object.keys(data).length === 0) {
      return { success: false, error: 'No data provided' };
    }

    try {
      const connector = ConnectorFactory.createConnector(connection.db_type);
      await connector.connect(connection);

      let result;
      if (connection.db_type === 'mongodb') {
        result = await this.insertMongoDB(connector, table, data);
      } else {
        result = await this.insertSQL(connector, table, schemaName, data);
      }

      await connector.disconnect();
      return result;
    } catch (error) {
      logger.error(`Insert row failed on '${connection.name}.${table}':`, error);
      return { success: false, error: error.message };
    }
  }

  async deleteRow(connection, table, schemaName, primaryKey) {
    if (!primaryKey || Object.keys(primaryKey).length === 0) {
      return { success: false, error: 'Primary key required for delete' };
    }

    try {
      const connector = ConnectorFactory.createConnector(connection.db_type);
      await connector.connect(connection);

      let result;
      if (connection.db_type === 'mongodb') {
        result = await this.deleteMongoDB(connector, table, primaryKey);
      } else {
        result = await this.deleteSQL(connector, table, schemaName, primaryKey);
      }

      await connector.disconnect();
      return result;
    } catch (error) {
      logger.error(`Delete row failed on '${connection.name}.${table}':`, error);
      return { success: false, error: error.message };
    }
  }

  async updateSQL(connector, table, schema, primaryKey, changes) {
    const setClause = Object.entries(changes)
      .map(([col, val]) => `"${col}" = '${val}'`)
      .join(', ');
    const whereClause = Object.entries(primaryKey)
      .map(([col, val]) => `"${col}" = '${val}'`)
      .join(' AND ');

    let query;
    if (schema && schema !== 'main' && schema !== 'public') {
      query = `UPDATE ${schema}."${table}" SET ${setClause} WHERE ${whereClause}`;
    } else {
      query = `UPDATE "${table}" SET ${setClause} WHERE ${whereClause}`;
    }

    const result = await connector.executeQuery(query);

    if (result.success) {
      return { success: true, message: 'Row updated successfully' };
    } else {
      return { success: false, error: result.error || 'Update failed' };
    }
  }

  async insertSQL(connector, table, schema, data) {
    const columns = Object.keys(data).map(col => `"${col}"`).join(', ');
    const values = Object.values(data).map(val => `'${val}'`).join(', ');

    let query;
    if (schema && schema !== 'main' && schema !== 'public') {
      query = `INSERT INTO ${schema}."${table}" (${columns}) VALUES (${values})`;
    } else {
      query = `INSERT INTO "${table}" (${columns}) VALUES (${values})`;
    }

    const result = await connector.executeQuery(query);

    if (result.success) {
      return { success: true, message: 'Row inserted successfully' };
    } else {
      return { success: false, error: result.error || 'Insert failed' };
    }
  }

  async deleteSQL(connector, table, schema, primaryKey) {
    const whereClause = Object.entries(primaryKey)
      .map(([col, val]) => `"${col}" = '${val}'`)
      .join(' AND ');

    let query;
    if (schema && schema !== 'main' && schema !== 'public') {
      query = `DELETE FROM ${schema}."${table}" WHERE ${whereClause}`;
    } else {
      query = `DELETE FROM "${table}" WHERE ${whereClause}`;
    }

    const result = await connector.executeQuery(query);

    if (result.success) {
      return { success: true, message: 'Row deleted successfully' };
    } else {
      return { success: false, error: result.error || 'Delete failed' };
    }
  }

  async updateMongoDB(connector, collection, filterDoc, changes) {
    try {
      const db = connector.connection.db();
      const coll = db.collection(collection);
      const result = await coll.updateOne(filterDoc, { $set: changes });

      if (result.modifiedCount > 0) {
        return { success: true, message: 'Document updated successfully' };
      } else {
        return { success: false, error: 'No document matched the filter' };
      }
    } catch (error) {
      return { success: false, error: error.message };
    }
  }

  async insertMongoDB(connector, collection, data) {
    try {
      const db = connector.connection.db();
      const coll = db.collection(collection);
      const result = await coll.insertOne(data);

      return {
        success: true,
        message: 'Document inserted successfully',
        id: result.insertedId.toString(),
      };
    } catch (error) {
      return { success: false, error: error.message };
    }
  }

  async deleteMongoDB(connector, collection, filterDoc) {
    try {
      const db = connector.connection.db();
      const coll = db.collection(collection);
      const result = await coll.deleteOne(filterDoc);

      if (result.deletedCount > 0) {
        return { success: true, message: 'Document deleted successfully' };
      } else {
        return { success: false, error: 'No document matched the filter' };
      }
    } catch (error) {
      return { success: false, error: error.message };
    }
  }
}

const dataEditor = new DataEditor();

module.exports = { DataEditor, dataEditor };
